--- a/net/minecraft/world/level/biome/BiomeSpecialEffects.java
+++ b/net/minecraft/world/level/biome/BiomeSpecialEffects.java
@@ -38,18 +_,18 @@
          return p_151760_.f_47938_;
       })).apply(p_47971_, BiomeSpecialEffects::new);
    });
-   private final int f_47927_;
-   private final int f_47928_;
-   private final int f_47929_;
-   private final int f_47930_;
-   private final Optional<Integer> f_47931_;
-   private final Optional<Integer> f_47932_;
-   private final BiomeSpecialEffects.GrassColorModifier f_47933_;
-   private final Optional<AmbientParticleSettings> f_47934_;
-   private final Optional<Holder<SoundEvent>> f_47935_;
-   private final Optional<AmbientMoodSettings> f_47936_;
-   private final Optional<AmbientAdditionsSettings> f_47937_;
-   private final Optional<Music> f_47938_;
+   public int f_47927_; // CatServer - private -> public  remove final  for Fabric BiomeModificationContextImpl
+   public int f_47928_; // CatServer - private -> public  remove final  for Fabric BiomeModificationContextImpl
+   public int f_47929_; // CatServer - private -> public  remove final  for Fabric BiomeModificationContextImpl
+   public int f_47930_; // CatServer - private -> public  remove final  for Fabric BiomeModificationContextImpl
+   public Optional<Integer> f_47931_; // CatServer - private -> public  remove final  for Fabric BiomeModificationContextImpl
+   public Optional<Integer> f_47932_; // CatServer - private -> public  remove final  for Fabric BiomeModificationContextImpl
+   public BiomeSpecialEffects.GrassColorModifier f_47933_; // CatServer - private -> public  remove final  for Fabric BiomeModificationContextImpl
+   public Optional<AmbientParticleSettings> f_47934_; // CatServer - private -> public  remove final  for Fabric BiomeModificationContextImpl
+   public Optional<Holder<SoundEvent>> f_47935_; // CatServer - private -> public  remove final  for Fabric BiomeModificationContextImpl
+   public Optional<AmbientMoodSettings> f_47936_; // CatServer - private -> public  remove final  for Fabric BiomeModificationContextImpl
+   public Optional<AmbientAdditionsSettings> f_47937_; // CatServer - private -> public  remove final  for Fabric BiomeModificationContextImpl
+   public Optional<Music> f_47938_; // CatServer - private -> public  remove final  for Fabric BiomeModificationContextImpl
 
    BiomeSpecialEffects(int p_47941_, int p_47942_, int p_47943_, int p_47944_, Optional<Integer> p_47945_, Optional<Integer> p_47946_, BiomeSpecialEffects.GrassColorModifier p_47947_, Optional<AmbientParticleSettings> p_47948_, Optional<Holder<SoundEvent>> p_47949_, Optional<AmbientMoodSettings> p_47950_, Optional<AmbientAdditionsSettings> p_47951_, Optional<Music> p_47952_) {
       this.f_47927_ = p_47941_;
@@ -201,7 +_,7 @@
       }
    }
 
-   public static enum GrassColorModifier implements StringRepresentable {
+   public static enum GrassColorModifier implements StringRepresentable, net.minecraftforge.common.IExtensibleEnum {
       NONE("none") {
          public int m_6583_(double p_48081_, double p_48082_, int p_48083_) {
             return p_48083_;
@@ -220,14 +_,37 @@
       };
 
       private final String f_48051_;
-      public static final Codec<BiomeSpecialEffects.GrassColorModifier> f_48050_ = StringRepresentable.m_216439_(BiomeSpecialEffects.GrassColorModifier::values);
+      public static final Codec<BiomeSpecialEffects.GrassColorModifier> f_48050_ = net.minecraftforge.common.IExtensibleEnum.createCodecForExtensibleEnum(BiomeSpecialEffects.GrassColorModifier::values, BiomeSpecialEffects.GrassColorModifier::byName);
+      private static final java.util.Map<String, GrassColorModifier> BY_NAME = java.util.Arrays.stream(values()).collect(java.util.stream.Collectors.toMap(BiomeSpecialEffects.GrassColorModifier::m_48072_, grassColorModifier -> grassColorModifier));
 
-      public abstract int m_6583_(double p_48065_, double p_48066_, int p_48067_);
+      public int m_6583_(double p_48065_, double p_48066_, int p_48067_) {
+         return delegate.modifyGrassColor(p_48065_, p_48066_, p_48067_);
+      }
 
       GrassColorModifier(String p_48058_) {
          this.f_48051_ = p_48058_;
       }
 
+      private ColorModifier delegate;
+      private GrassColorModifier(String name, ColorModifier delegate) {
+         this(name);
+         this.delegate = delegate;
+      }
+      public static GrassColorModifier create(String name, String id, ColorModifier delegate) {
+         throw new IllegalStateException("Enum not extended");
+      }
+      @Override
+      public void init() {
+         BY_NAME.put(this.m_48072_(), this);
+      }
+      // Forge: Access enum members by name
+      public static BiomeSpecialEffects.GrassColorModifier byName(String name) {
+         return BY_NAME.get(name);
+      }
+      @FunctionalInterface
+      public interface ColorModifier {
+         int modifyGrassColor(double x, double z, int color);
+      }
       public String m_48072_() {
          return this.f_48051_;
       }
