--- a/net/minecraft/world/MenuProvider.java
+++ b/net/minecraft/world/MenuProvider.java
@@ -3,6 +_,14 @@
 import net.minecraft.network.chat.Component;
 import net.minecraft.world.inventory.MenuConstructor;
 
-public interface MenuProvider extends MenuConstructor {
+public interface MenuProvider extends MenuConstructor, net.fabricmc.fabric.api.screenhandler.v1.FabricScreenHandlerFactory { // Fabric - NamedScreenHandlerFactoryMixin
    Component m_5446_();
+
+   // CatServer start
+   net.minecraft.core.BlockPos[] blockPosForEvent = new net.minecraft.core.BlockPos[1];
+   default MenuProvider setBlockPosForEvent(net.minecraft.core.BlockPos pos) {
+      this.blockPosForEvent[0] = pos;
+      return this;
+   }
+   // CatServer end
 }
