--- a/net/minecraft/world/SimpleMenuProvider.java
+++ b/net/minecraft/world/SimpleMenuProvider.java
@@ -8,7 +_,7 @@
 
 public final class SimpleMenuProvider implements MenuProvider {
    private final Component f_19199_;
-   private final MenuConstructor f_19200_;
+   public final MenuConstructor f_19200_; // CatServer - private -> public for Fabric
 
    public SimpleMenuProvider(MenuConstructor p_19202_, Component p_19203_) {
       this.f_19200_ = p_19202_;
