--- a/net/minecraft/world/item/ShovelItem.java
+++ b/net/minecraft/world/item/ShovelItem.java
@@ -18,8 +_,8 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.gameevent.GameEvent;
 
-public class ShovelItem extends DiggerItem {
-   protected static final Map<Block, BlockState> f_43110_ = Maps.newHashMap((new ImmutableMap.Builder()).put(Blocks.f_50440_, Blocks.f_152481_.m_49966_()).put(Blocks.f_50493_, Blocks.f_152481_.m_49966_()).put(Blocks.f_50599_, Blocks.f_152481_.m_49966_()).put(Blocks.f_50546_, Blocks.f_152481_.m_49966_()).put(Blocks.f_50195_, Blocks.f_152481_.m_49966_()).put(Blocks.f_152549_, Blocks.f_152481_.m_49966_()).build());
+public class ShovelItem extends DiggerItem implements net.fabricmc.fabric.mixin.content.registry.ShovelItemAccessor { // Fabric - ShovelItemAccessor
+   public static final Map<Block, BlockState> f_43110_ = Maps.newHashMap((new ImmutableMap.Builder()).put(Blocks.f_50440_, Blocks.f_152481_.m_49966_()).put(Blocks.f_50493_, Blocks.f_152481_.m_49966_()).put(Blocks.f_50599_, Blocks.f_152481_.m_49966_()).put(Blocks.f_50546_, Blocks.f_152481_.m_49966_()).put(Blocks.f_50195_, Blocks.f_152481_.m_49966_()).put(Blocks.f_152549_, Blocks.f_152481_.m_49966_()).build()); // CatServer - protected -> public for Fabric ShovelItemAccessor
 
    public ShovelItem(Tier p_43114_, float p_43115_, float p_43116_, Item.Properties p_43117_) {
       super(p_43115_, p_43116_, p_43114_, BlockTags.f_144283_, p_43117_);
@@ -33,9 +_,9 @@
          return InteractionResult.PASS;
       } else {
          Player player = p_43119_.m_43723_();
-         BlockState blockstate1 = f_43110_.get(blockstate.m_60734_());
+         BlockState blockstate1 = blockstate.getToolModifiedState(p_43119_, net.minecraftforge.common.ToolActions.SHOVEL_FLATTEN, false);
          BlockState blockstate2 = null;
-         if (blockstate1 != null && level.m_8055_(blockpos.m_7494_()).m_60795_()) {
+         if (blockstate1 != null && level.m_46859_(blockpos.m_7494_())) {
             level.m_5594_(player, blockpos, SoundEvents.f_12406_, SoundSource.BLOCKS, 1.0F, 1.0F);
             blockstate2 = blockstate1;
          } else if (blockstate.m_60734_() instanceof CampfireBlock && blockstate.m_61143_(CampfireBlock.f_51227_)) {
@@ -63,5 +_,15 @@
             return InteractionResult.PASS;
          }
       }
+   }
+
+   @org.jetbrains.annotations.Nullable
+   public static BlockState getShovelPathingState(BlockState originalState) {
+      return f_43110_.get(originalState.m_60734_());
+   }
+
+   @Override
+   public boolean canPerformAction(ItemStack stack, net.minecraftforge.common.ToolAction toolAction) {
+      return net.minecraftforge.common.ToolActions.DEFAULT_SHOVEL_ACTIONS.contains(toolAction);
    }
 }
