--- a/net/minecraft/server/dedicated/DedicatedServerProperties.java
+++ b/net/minecraft/server/dedicated/DedicatedServerProperties.java
@@ -43,6 +_,7 @@
 import org.slf4j.Logger;
 
 public class DedicatedServerProperties extends Settings<DedicatedServerProperties> {
+   public final boolean debug = this.m_139836_("debug", false); // CraftBukkit
    static final Logger f_214805_ = LogUtils.getLogger();
    private static final Pattern f_214806_ = Pattern.compile("^[a-fA-F0-9]{40}$");
    private static final Splitter f_243988_ = Splitter.on(',').trimResults();
@@ -115,7 +_,7 @@
          return p_211541_.toLowerCase(Locale.ROOT);
       }, WorldPresets.f_226437_.m_135782_().toString()));
       this.f_214802_ = m_214808_(this.m_139811_("resource-pack", ""), this.m_139811_("resource-pack-sha1", ""), this.m_139803_("resource-pack-hash"), this.m_139836_("require-resource-pack", false), this.m_139811_("resource-pack-prompt", ""));
-      this.f_243997_ = m_247013_(this.m_139811_("initial-enabled-packs", String.join(",", WorldDataConfiguration.f_244649_.f_244096_().m_45850_())), this.m_139811_("initial-disabled-packs", String.join(",", WorldDataConfiguration.f_244649_.f_244096_().m_45855_())));
+      this.f_243997_ = m_247013_(this.m_139811_("initial-enabled-packs", String.join(",", net.fabricmc.fabric.impl.resource.loader.ModResourcePackUtil.createDefaultDataConfiguration().f_244096_().m_45850_())), this.m_139811_("initial-disabled-packs", String.join(",", WorldDataConfiguration.f_244649_.f_244096_().m_45855_()))); // Fabric - ServerPropertiesHandlerMixin
    }
 
    public static DedicatedServerProperties m_180929_(Path p_180930_) {
@@ -189,7 +_,7 @@
       return this.f_244504_.m_247373_(p_250466_);
    }
 
-   static record WorldDimensionData(JsonObject f_244404_, String f_243780_) {
+   public static record WorldDimensionData(JsonObject f_244404_, String f_243780_) { // CatServer - public
       private static final Map<String, ResourceKey<WorldPreset>> f_244358_ = Map.of("default", WorldPresets.f_226437_, "largebiomes", WorldPresets.f_226439_);
 
       public WorldDimensions m_247373_(RegistryAccess p_248812_) {
