--- a/net/minecraft/server/commands/GameRuleCommand.java
+++ b/net/minecraft/server/commands/GameRuleCommand.java
@@ -8,13 +_,23 @@
 import net.minecraft.network.chat.Component;
 import net.minecraft.world.level.GameRules;
 
-public class GameRuleCommand {
+public class GameRuleCommand implements net.fabricmc.fabric.mixin.gamerule.GameRuleCommandAccessor { // Fabric - GameRuleCommandAccessor
    public static void m_137744_(CommandDispatcher<CommandSourceStack> p_137745_) {
       final LiteralArgumentBuilder<CommandSourceStack> literalargumentbuilder = Commands.m_82127_("gamerule").requires((p_137750_) -> {
          return p_137750_.m_6761_(2);
       });
       GameRules.m_46164_(new GameRules.GameRuleTypeVisitor() {
          public <T extends GameRules.Value<T>> void m_6889_(GameRules.Key<T> p_137764_, GameRules.Type<T> p_137765_) {
+            {
+               // Fabric start - GameRuleCommandVisitorMixin
+               // Check if our type is a EnumRuleType
+               if (p_137765_ instanceof net.fabricmc.fabric.impl.gamerule.EnumRuleType) {
+                  //noinspection rawtypes,unchecked
+                  net.fabricmc.fabric.impl.gamerule.EnumRuleCommand.register(literalargumentbuilder, (GameRules.Key)p_137764_, (net.fabricmc.fabric.impl.gamerule.EnumRuleType) p_137765_);
+                  return;
+               }
+               // Fabric end
+            }
             literalargumentbuilder.then(Commands.m_82127_(p_137764_.m_46328_()).executes((p_137771_) -> {
                return GameRuleCommand.m_137757_(p_137771_.getSource(), p_137764_);
             }).then(p_137765_.m_46358_("value").executes((p_137768_) -> {
@@ -25,9 +_,9 @@
       p_137745_.register(literalargumentbuilder);
    }
 
-   static <T extends GameRules.Value<T>> int m_137754_(CommandContext<CommandSourceStack> p_137755_, GameRules.Key<T> p_137756_) {
+   public static <T extends GameRules.Value<T>> int m_137754_(CommandContext<CommandSourceStack> p_137755_, GameRules.Key<T> p_137756_) { // CatServer - public for Fabric GameRuleCommandAccessor
       CommandSourceStack commandsourcestack = p_137755_.getSource();
-      T t = commandsourcestack.m_81377_().m_129900_().m_46170_(p_137756_);
+      T t = commandsourcestack.m_81372_().m_46469_().m_46170_(p_137756_); // CraftBukkit
       t.m_46370_(p_137755_, "value");
       commandsourcestack.m_288197_(() -> {
          return Component.m_237110_("commands.gamerule.set", p_137756_.m_46328_(), t.toString());
@@ -35,8 +_,8 @@
       return t.m_6855_();
    }
 
-   static <T extends GameRules.Value<T>> int m_137757_(CommandSourceStack p_137758_, GameRules.Key<T> p_137759_) {
-      T t = p_137758_.m_81377_().m_129900_().m_46170_(p_137759_);
+   public static <T extends GameRules.Value<T>> int m_137757_(CommandSourceStack p_137758_, GameRules.Key<T> p_137759_) { // CatServer - public for Fabric GameRuleCommandAccessor
+      T t = p_137758_.m_81372_().m_46469_().m_46170_(p_137759_); // CraftBukkit
       p_137758_.m_288197_(() -> {
          return Component.m_237110_("commands.gamerule.query", p_137759_.m_46328_(), t.toString());
       }, false);
